runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -9317,6 +9317,391 @@ static TR::Register* inlineStringHashCode(TR::Node* node, bool isCompressed, TR:
+
+   OMR::X86::Encoding vectorEncoding = vectorEncodingMethods[vl - TR::VectorLength128];
+   int32_t vectorSizeElements = vectorSizes[vl - TR::VectorLength128];
+   int32_t numElements = vectorSizeElements * unrollCount;
+
+   TR::RegisterDependencyConditions *deps = generateRegisterDependencyConditions((uint8_t)0, (uint8_t)11, cg);
+   TR::Register *tmp = cg->allocateRegister(TR_GPR);
+   TR::Register *tmpXMM = cg->allocateRegister(TR_VRF);
+   TR::Register *multiplierXMM = cg->allocateRegister(TR_VRF);
+
+   TR::Register *hashXMM[]          = { cg->allocateRegister(TR_VRF), cg->allocateRegister(TR_VRF), cg->allocateRegister(TR_VRF), cg->allocateRegister(TR_VRF) };
+   TR::Register *multiplier31PowN[] = { cg->allocateRegister(TR_VRF), cg->allocateRegister(TR_VRF), cg->allocateRegister(TR_VRF), cg->allocateRegister(TR_VRF) };
+
+   deps->addPostCondition(tmp, TR::RealRegister::NoReg, cg);
+   deps->addPostCondition(tmpXMM, TR::RealRegister::NoReg, cg);
+   deps->addPostCondition(multiplierXMM, TR::RealRegister::NoReg, cg);
+
+   for (int32_t i = 0; i < 4; i++) deps->addPostCondition(multiplier31PowN[i], TR::RealRegister::NoReg, cg);
+   for (int32_t i = 0; i < 4; i++) deps->addPostCondition(hashXMM[i], TR::RealRegister::NoReg, cg);
+
+   TR::LabelSymbol *begLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *endLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *loopLabel = generateLabelSymbol(cg);
+
+   begLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+   generateRegRegInstruction(TR::InstOpCode::MOVRegReg(), node, result, initialHash, cg);
+   generateLabelInstruction(TR::InstOpCode::label, node, begLabel, cg);
+   generateRegRegInstruction(TR::InstOpCode::MOVRegReg(), node, tmp, length, cg);
+   generateRegImmInstruction(TR::InstOpCode::AND4RegImm4, node, tmp, ~(numElements - 1), cg);
+
+   {
+   generateRegRegInstruction(TR::InstOpCode::CMP4RegReg, node, index, tmp, cg);
+   generateLabelInstruction(TR::InstOpCode::JGE4, node, endLabel, cg);
+
+   // Initialize Constants outside of loop body but after the first compare
+   generateRegRegInstruction(TR::InstOpCode::MOVDRegReg4, node, hashXMM[0], initialHash, cg);
+   for (int32_t i = 1; i < unrollCount; i++)
+      generateRegRegInstruction(TR::InstOpCode::PXORRegReg, node, hashXMM[i], hashXMM[i], cg, vectorEncoding);
+
+   int32_t multiplier31PowNData[16];
+   // Fill multiplier array with 31^numElements
+   std::fill_n(multiplier31PowNData, 16, powersOf31[64 - numElements]);
+   generateRegMemInstruction(TR::InstOpCode::MOVDQURegMem, node, multiplierXMM, generateX86MemoryReference(cg->findOrCreateConstantDataSnippet(node, multiplier31PowNData, vectorSizeElements * sizeof(int32_t)), cg), cg, vectorEncoding);
+
+   for (int32_t i = 0; i < unrollCount; i++)
+      {
+      // Based on the unrolling factor (x), we need to split multiplier constant (powers of 31) into x vectors.
+      //   The constant is as follows (31^n-1, 31^n-2, ..., 31^0)
```

Comment:
Minor nit on the operator precedence to improve readability: `31^(n-1), 31^(n-2), ...`